github-actions/start-release/start_release.py

```diff
@@ -0,0 +1,164 @@
+import argparse
+import base64
+import datetime
+import json
+import logging
+import os
+import re
+from distutils.version import LooseVersion
+
+from api.github import GitHubApiSession
+
+RELEASE_NOTES_DIR = 'release_notes'
+UNRELEASED_MD = '{}/unreleased.md'.format(RELEASE_NOTES_DIR)
+UNRELEASED_MD_HEADER = '**Unreleased**'
+RELEASE_NOTES_MD = RELEASE_NOTES_DIR + '/{}.md'
+RELEASE_NOTES_TOP_HEADER = '**{} Release Notes - Published by {} {}**\n\n'
+RELEASE_NOTES_VERSION_HEADER = '**Version {} - Released <INSERT_RELEASE_DATE>**\n'
+RELEASE_NOTES_DATE_FORMAT = '%B %d, %Y'
+RELEASE_NOTE_PATTERN = re.compile('^\\* .+$')
+
+DEFAULT_ENCODING = 'utf-8'
+
+RELEASE_NOTE_COMMIT_MSG = 'Release notes for version {}'
+MAIN_PR_TITLE = 'Merging next to main for release {}'
+MAIN_PR_BODY_FILE = 'templates/main_pr_body.md'
+COMMIT_AUTHOR = {
+    'name': 'root',
+    'email': 'root@splunksoar'
+}
+
+
+def deserialize_app_json(app_json):
+    json_str = base64.b64decode(app_json['content']).decode(DEFAULT_ENCODING)
+    start = idx = json_str.find('\n') + 2
+    while json_str[idx] != '"':
+        idx += 1
+    indent = idx - start + 1
+
+    return json.loads(json_str), indent
+
+
+def deserialize_unreleased_md(unreleased_md):
+    return base64.b64decode(unreleased_md['content'])\
+        .decode(DEFAULT_ENCODING).split('\n')
+
+
+def build_release_notes(release_version, unreleased_notes, app_json):
+    if unreleased_notes[0] != UNRELEASED_MD_HEADER:
+        raise ValueError('Expected the first line of {} to be the header {}'
+                         .format(UNRELEASED_MD, UNRELEASED_MD_HEADER))
+    elif len(unreleased_notes) == 1:
+        return None
+
+    publish_date = datetime.date.today().strftime(RELEASE_NOTES_DATE_FORMAT)
+    release_notes = [
+        RELEASE_NOTES_TOP_HEADER.format(app_json['name'], app_json['publisher'], publish_date),
+        RELEASE_NOTES_VERSION_HEADER.format(release_version)
+    ]
+    for note in unreleased_notes[1:]:
+        if not note or not note.strip():
+            continue
+        match = RELEASE_NOTE_PATTERN.match(note)
+        if not match:
+            raise ValueError('Detected incorrectly formatted release note: \'{}\''
+                             .format(note))
+        release_notes.append(match.group())
+
+    return '\n'.join(release_notes)
+
+
+def load_main_pr_body():
+    with open(MAIN_PR_BODY_FILE) as f:
+        return f.read()
+
+
+def start_release(session):
+    # Fetch repo contents and find the app json file - which we
+    # expect to be the only json file
+    repo_files = session.get('contents?ref={}'.format('next'))
+    json_files = [f['name'] for f in repo_files if f['name'].lower().endswith('.json')]
+
+    if len(json_files) != 1:
+        raise ValueError('Expected a single json file in the repo but got {}'.format(json_files))
+
+    # Fetch the app json from main branch, and compare the app version
```

Comment:
Potentially naiive question - how is this going to work if the version we're about to release is the first version of an app?

Reply:
good point, if this is the first version then ```main``` would probably be empty

I think in this case, we can use a default first version like `1.0.0` to compare against?

Reply:
That seems reasonable to me